package-res/resources/web/pentaho/data/filter.js

```diff
@@ -0,0 +1,165 @@
+    }
+  });
+
+  /**
+   * This namespace contains classes for filters that are used for expressing a data set intensionally.
+   * It also exports a static create function that can be used to generate a filter tree from a spec.
+   *
+   * @name pentaho.data.filter
+   * @namespace
+   * @memberOf pentaho.data
+   * @amd pentaho/data/filter
+   *
+   * @example
+   * <caption> Two methods for creating a filter.</caption>
+   *
+   * require(["pentaho/data/Table", "pentaho/data/Filter"], function(Table, Filter) {
+   *   var data = new Table({
+   *     model: [
+   *       {name: "product", type: "string", label: "Product"},
+   *       {name: "sales", type: "number", label: "Sales"},
+   *       {name: "inStock", type: "boolean", label: "In Stock"}
+   *     ],
+   *     rows: [
+   *       {c: [{v: "A"}, {v: 12000}, {v: true}]},
+   *       {c: [{v: "B"}, {v: 6000}, {v: true}]},
+   *       {c: [{v: "C"}, {v: 12000}, {v: false}]},
+   *       {c: [{v: "D"}, {v: 1000}, {v: false}]},
+   *       {c: [{v: "E"}, {v: 2000}, {v: false}]},
+   *       {c: [{v: "F"}, {v: 3000}, {v: false}]},
+   *       {c: [{v: "G"}, {v: 4000}, {v: false}]}
+   *     ]
+   *   });
+   *
+   *   var sales12k = new Filter.IsEqual("sales", [12000]);
+   *   var productAB = new Filter.IsIn("product", ["A", "B"]);
+   *   var notInProductABFilter = new Not(productAB);
+   *   var andFilter = new Filter.And([sales12k, notInProductABFilter]);
+   *   var filter = new Not(andFilter);
+   *   var filteredData = filter.apply(data);
+   *   // filteredData.getValue(0, 0) === "A"
+   *   // filteredData.getValue(1, 0) === "B"
+   *   // filteredData.getValue(2, 0) === "D"
+   *   // filteredData.getValue(3, 0) === "E"
+   *   // filteredData.getValue(4, 0) === "F"
+   *   // filteredData.getValue(5, 0) === "G"
+   *
+   *   //Or alternatively,
+   *   var spec = {
+   *      "$not": {
+   *        "$and": [
+   *          {"sales": 12000},
+   *          {"$not": {"product": {"$in": ["A", "B"]}}}
+   *        ]
+   *      }
+   *    };
+   *
+   *   var filterFromSpec = Filter.create(spec);
+   *   var filteredDataFromSpec = filterFromSpec.apply(data);
+   *   // filteredDataFromSpec.getValue(0, 0) === "A"
+   *   // filteredDataFromSpec.getValue(1, 0) === "B"
+   *   // filteredDataFromSpec.getValue(2, 0) === "D"
+   *   // filteredDataFromSpec.getValue(3, 0) === "E"
+   *   // filteredDataFromSpec.getValue(4, 0) === "F"
+   *   // filteredDataFromSpec.getValue(5, 0) === "G"
+   * });
+   *
+   */
+  return {
+    // Leaf nodes
+    IsEqual: IsEqual,
+    IsIn: IsIn,
+    // Non-leaf nodes
+    Or: OrFilter,
+    And: AndFilter,
+    Not: NotFilter,
+    //Root: RootFilter,
+
+    /**
+     * Create a filter from a spec
+     *
+     * @memberOf pentaho.data.filter
+     * @method
+     * @static
+     * @param {Object} spec - Specification of a filter
+     * @return {pentaho.data.filter.AbstractFilter} The tree of filter objects that corresponds to the specification
```

Comment:
No, we should just use "!Class" or "?string" to change default
interpretation. Classes are nullable by default. Non-classes are not
nullable by default.